### PR TITLE
34 middleware for express routes

### PR DESCRIPTION
**Pull request by erensunerr**
- Created middlewares for error handling and logging.
- Refactored the api routes & controllers.

IMPORTANT: the message of any error raised in a controller will be returned in the error handler response.

### CODE REVIEW

File: server/controllers/ChallengeController.js [hunk 1]
```diff
@@ -2,119 +2,94 @@ const db = require("../models/index");
...
-    }
+  switch (difficulty) {
+    case "easy":
+      return 0;
+    case "medium":
+      return 1;
+    case "hard":
+      return 2;
+    default:
+      return -1;
```
Comment (alexapostolu) [inline server/controllers/ChallengeController.js:13]: Good job for keeping the indentation consistent!

File: server/controllers/ChallengeController.js [hunk 2]
```diff
@@ -2,119 +2,94 @@ const db = require("../models/index");
...
-                expectedFunctionality: challengeDescription.expectedFunctionality
-            }
-        });
-
-        res.status(200).json(challenges);
-    }
-    catch (error) {
-        res.status(500).json({ error: error.message });
-    }
+  const challengesData = await Challenge.findAll();
```
Comment (alexapostolu) [inline server/controllers/ChallengeController.js:18]: Why did you remove the try-catch block here?

File: server/controllers/ChallengeController.js [hunk 3]
```diff
@@ -2,119 +2,94 @@ const db = require("../models/index");
...
-            expectedFunctionality: challengeDescription.expectedFunctionality
-        }
-
-        res.status(200).json(challenge);
-    }
-	catch (error) {
-        res.status(500).json({ error: error.message });
-    }
-}
+  const challenge_id = req.params.id;
```
Comment (alexapostolu) [inline server/controllers/ChallengeController.js:37]: Same here, why remove the try-catch?

File: server/routes/ErrorHandlingMiddleware.js [hunk 4]
```diff
@@ -0,0 +1,23 @@
+function ErrorHandler(err, req, res, next) {
+  console.error("Error", err);
+  res.status(500).json({
```
Comment (o-bm) [inline server/routes/ErrorHandlingMiddleware.js:3]: I think this function will always set 500 status error which indicates server error but there can be other errors too like 404 or 401

File: server/routes/LoggingMiddleware.js [hunk 5]
```diff
@@ -0,0 +1,6 @@
+function LoggingMiddleware(req, res, next) {
```
Comment (o-bm) [inline server/routes/LoggingMiddleware.js:1]: What does this do ? Like are we just calling 'next()' for now temporarily as we are not fully done with auth?

File: server/controllers/ChallengeController.js (hunk 3 above)
Comment (erensunerr) [reply server/controllers/ChallengeController.js:37]: Hey Alex, the try-catch blocks were really repetitive so I extracted them into a middleware to do the error handling. They are all the same code, so why repeat it for every route?

File: server/routes/ErrorHandlingMiddleware.js (hunk 4 above)
Comment (erensunerr) [reply server/routes/ErrorHandlingMiddleware.js:3]: Great catch! I created a HandledError as a subclass of Error so we can pass error codes.

File: server/routes/LoggingMiddleware.js (hunk 5 above)
Comment (erensunerr) [reply server/routes/LoggingMiddleware.js:1]: No, this just logs all the requests to the console. When deploying we can redirect the output to a file or add the option to write logs to a file.

**Review by arthurhohinng — APPROVED** [review]
Looks good to me
